File: src/epcap.erl
%%% @copyright 2009-2023 Michael Santos <michael.santos@gmail.com>
%%% All rights reserved.
%%%
%%% Redistribution and use in source and binary forms, with or without
%%% modification, are permitted provided that the following conditions
%%% are met:
%%%
%%% 1. Redistributions of source code must retain the above copyright notice,
%%% this list of conditions and the following disclaimer.
%%%
%%% 2. Redistributions in binary form must reproduce the above copyright
%%% notice, this list of conditions and the following disclaimer in the
%%% documentation and/or other materials provided with the distribution.
%%%
%%% 3. Neither the name of the copyright holder nor the names of its
%%% contributors may be used to endorse or promote products derived from
%%% this software without specific prior written permission.
%%%
%%% THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
%%% "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
%%% LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
%%% A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
%%% HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
%%% SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
%%% TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
%%% PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
%%% LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
%%% NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
%%% SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

%% @doc An Erlang port interface to libpcap.
%%
%% == SETUP ==
%%
%% ```
%% # Allow your user to epcap with root privs
%% sudo visudo
%% youruser ALL = NOPASSWD: /path/to/epcap/priv/epcap
%% # And if requiretty is enabled, disable it by using one of these
%% Defaults!/path/to/epcap/priv/epcap !requiretty
%% Defaults:youruser !requiretty
%% '''
%%
%% == EXAMPLES ==
%%
%% To run the `sniff' example:
%%
%% ```
%% $ mkdir -p ebin
%% $ erlc +debug_info -I _build/default/lib -o ebin examples/*.erl
%% '''
%%
%% ```
%% rebar3 shell
%%
%% % Start the sniffer process
%% sniff:start_link().
%%
%% % Use your interface, or leave it out and trust in pcap
%% sniff:start([{interface, "eth0"}]).
%%
%% % To change the filter
%% sniff:start([{filter, "icmp or (tcp and port 80)"},{interface, "eth0"}]).
%%
%% % To stop sniffing
%% sniff:stop().
%% '''
%%
%% == PF_RING ==
%%
%% In case you want to compile epcap with PF_RING support, just specify
%% the path to the libpfring and modified libpcap libraries via shell
%% variable PFRING.
%%
%% ```
%% PFRING=/home/user/pfring rebar3 do clean, compile
%% '''
%%
%% To complete the configuration you need to set up the cluster_id
%% option. The value of the cluster_id option is integer and should be in
%% range between 0 and 255.
%%
%% ```
%% epcap:start_link([{interface, "lo"}, {cluster_id, 2}]).
%% '''
%%
%% You can also specify the option cpu_affinity to set up CPU affinity for
%% epcap port:
%%
%% ```
%% epcap:start_link([{interface, "lo"}, {cluster_id, 2}, {cpu_affinity, "1,3,5-7"}]).
%% '''
%%
%% == PROCESS RESTRICTIONS ==
%%
%% Setting the `RESTRICT_PROCESS' environment variable controls which
%% mode of process restriction is used. The available modes are:
%% 
%% * seccomp: linux
%% 
%% * pledge: openbsd (default)
%% 
%% * capsicum: freebsd (default)
%% 
%% * rlimit: all (default: linux)
%% 
%% * null: all
%% 
%% For example, to force using the seccomp process restriction on linux:
%% 
%% ```
%% RESTRICT_PROCESS=rlimit rebar3 do clean, compile
%% '''
%% 
%% The `null' mode disables process restrictions and can be used for debugging.
%% 
%% ```
%% RESTRICT_PROCESS=null rebar3 do clean, compile
%% 
%% epcap:start([{exec, "sudo strace -f -s 4096 -o rlimit.trace"}, {filter, "port 9997"}]).
%% 
%% RESTRICT_PROCESS=seccomp rebar3 do clean, compile
%% 
%% epcap:start([{exec, "sudo strace -f -s 4096 -o seccomp.trace"}, {filter, "port 9997"}]).
%% '''
-module(epcap).

-behaviour(gen_server).

%% API
-export([start_link/0, start_link/1, start_link/2, stop/1]).

-export([start/0, start/1, start/2]).

-export([send/2]).

-export([getopts/1]).

%% gen_server callbacks
-export([
    init/1,
    handle_call/3,
    handle_cast/2,
    handle_info/2,
    terminate/2,
    code_change/3
]).

-export_type([time_unit/0, options/0]).

-record(state, {pid :: pid(), port :: port()}).

-type time_unit() :: timestamp | microsecond.

%% @doc Start the epcap port process.
%% @see start_link/2
-spec start() -> ignore | {error, _} | {ok, pid()}.
start() -> start(self(), []).

%% @doc Start the epcap port process with options.
%% @see start_link/2
-spec start(options()) -> ignore | {error, _} | {ok, pid()}.
start(Options) -> start(self(), Options).

%% @doc Start the epcap port process with owner and options.
%% @see start_link/2
-spec start(pid(), options()) -> ignore | {error, _} | {ok, pid()}.
start(Pid, Options) when is_pid(Pid), is_list(Options) ->
    gen_server:start(?MODULE, [Pid, Options], []).

%% @doc Start and link with the epcap port process.
%% @see start_link/2
-spec start_link() -> ignore | {error, _} | {ok, pid()}.
start_link() -> start_link(self(), []).

%% @doc Start and link with the epcap port process.
%% @see start_link/2
-spec start_link(options()) -> ignore | {error, _} | {ok, pid()}.
start_link(Options) -> start_link(self(), Options).

%% @doc Start and link with the epcap port process.
%%
%% Packets are delivered as messages:
%%
%% ```
%% {packet, DataLinkType, Time, Length, Packet}
%% '''
%%
%% The DataLinkType is an integer representing the link layer,
%% e.g., ethernet, Linux cooked socket.
%%
%% The Time can be either in microseconds or a timestamp in the same
%% format as erlang:now/0 depending on the value of the time_unit
%% option (default: timestamp):
%%
%% ```
%% {MegaSecs, Secs, MicroSecs}
%% '''
%%
%% The Length corresponds to the actual packet length on the
%% wire. The captured packet may have been truncated. To get the
%% captured packet length, use byte_size(Packet).
%%
%% The Packet is a binary holding the captured data.
%%
%% If the version of the pcap library supports it, the pcap buffer
%% size can be set to avoid dropped packets by using the 'buffer'
%% option. The buffer size must be larger than the snapshot
%% length (default: 65535) plus some overhead for the pcap data
%% structures. Using some multiple of the snapshot length is
%% suggested.
%%
%% == Examples ==
%%
%% ```
%% 1> epcap:start([{filter, "icmp or (tcp and port 80)"},{interface, "eth0"}]).
%% {ok,<0.197.0>}
%% 2> flush().
%% Shell got {packet,1,
%%                   {1693,57279,417781},
%%                   98,
%%                   <<0,22,62,179,157,38,0,22,62,91,72,102,8,0,69,0,0,84,231,22,
%%                     64,0,64,1,130,73,100,115,92,198,8,8,8,8,8,0,164,231,23,
%%                     127,0,1,255,0,234,100,0,0,0,0,141,95,6,0,0,0,0,0,16,17,18,
%%                     19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,
%%                     38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55>>}
%% Shell got {packet,1,
%%                   {1693,57279,450447},
%%                   98,
%%                   <<0,22,62,91,72,102,0,22,62,179,157,38,8,0,69,0,0,84,0,0,0,
%%                     0,115,1,118,96,8,8,8,8,100,115,92,198,0,0,172,231,23,127,
%%                     0,1,255,0,234,100,0,0,0,0,141,95,6,0,0,0,0,0,16,17,18,19,
%%                     20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,
%%                     39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55>>}
%% ok
%% '''
-spec start_link(pid(), options()) -> ignore | {error, _} | {ok, pid()}.
start_link(Pid, Options) -> gen_server:start_link(?MODULE, [Pid, Options], []).

%% @doc Inject a packet on the network interface.
%%
%% To enable sending packets, start_link/1 must be called with the
%% `{inject, true}' option (default: `{inject, false}'). When disabled,
%% any data sent to the epcap port is silently discarded. Packet injection
%% failures are treated as fatal errors terminating the epcap port. Partial
%% writes are not considered to be errors and are ignored (an error message
%% will be printed to stderr if the verbose option is used).
%%
%% == Examples ==
%%
%% ```
%% 1> {ok, Port} = epcap:start([inject]).
%% {ok,<0.197.0>}
%% 2> {packet, _, _, _, Packet} = receive N -> N end.
%% {packet,1,
%%         {1693,58033,222196},
%%         74,
%%         <<0,22,62,179,157,38,0,22,62,91,72,102,8,0,69,0,0,60,205,
%%           40,64,0,64,6,...>>}
%% 3> epcap:send(Port, Packet).
%% ok
%% '''
-spec send(pid(), iodata()) -> ok.
send(Pid, Packet) when is_pid(Pid) ->
    case iolist_size(Packet) < 65535 of
        true -> gen_server:call(Pid, {send, Packet}, infinity);
        false -> erlang:error(badarg)
    end.

%% @doc Stop the epcap port.
%%
%% == Examples ==
%%
%% ```
%% 1> {ok, Port} = epcap:start([{filter, "icmp or (tcp and port 80)"},{interface, "eth0"}]).
%% {ok,<0.205.0>}
%% 2> epcap:stop(Port).
%% ok
%% '''
-spec stop(pid()) -> ok.
stop(Pid) ->
    catch gen_server:call(Pid, stop),
    ok.

%% @private
init([Pid, Options0]) ->
    process_flag(trap_exit, true),
    Options = setopts(
        [
            {chroot, filename:join([basedir(), "tmp"])},
            {timeout, timeout()},
            {direction, inout}
        ],
        Options0
    ),
    ok = filelib:ensure_dir(
        filename:join(
            proplists:get_value(chroot, Options),
            "."
        )
    ),
    [Cmd | Argv] = getopts(Options),
    Port = open_port(
        {spawn_executable, Cmd},
        [{args, Argv}, {packet, 2}, binary, exit_status]
    ),
    % Block until the port has fully initialized
    receive
        {Port, {data, Data}} ->
            {epcap, ready} = binary_to_term(Data),
            {ok, #state{pid = Pid, port = Port}};
        {'EXIT', Port, normal} ->
            {stop, {error, port_init_failed}};
        {'EXIT', Port, Reason} ->
            {stop, {error, Reason}}
    end.

%% @private
handle_call({send, Packet}, _From, #state{port = Port} = State) ->
    Reply =
        try erlang:port_command(Port, Packet) of
            true -> ok
        catch
            error:badarg -> {error, closed}
        end,
    {reply, Reply, State};
handle_call(stop, _From, State) ->
    {stop, normal, ok, State}.

%% @private
handle_cast(_Msg, State) -> {noreply, State}.

%% @private
terminate(_Reason, #state{port = Port}) ->
    catch erlang:port_close(Port),
    ok.

%% @private
code_change(_OldVsn, State, _Extra) -> {ok, State}.

%%--------------------------------------------------------------------
%%% Port communication
%%--------------------------------------------------------------------

%% @private
handle_info({Port, {data, Data}}, #state{port = Port, pid = Pid} = State) ->
    Pid ! binary_to_term(Data),
    {noreply, State};
handle_info({'EXIT', Port, Reason}, #state{port = Port} = State) ->
    {stop, {shutdown, Reason}, State};
handle_info({Port, {exit_status, Status}}, #state{port = Port} = State) ->
    case Status of
        0 -> {stop, normal, State};
        _ when 128 < Status -> {stop, {port_terminated, Status - 128}, State};
        _ -> {stop, {port_terminated, Status}, State}
    end;
handle_info(
    Info,
    %% WTF
    State
) ->
    error_logger:error_report([{wtf, Info}]),
    {noreply, State}.

%%--------------------------------------------------------------------
%%% Internal functions
%%--------------------------------------------------------------------
-type arg_num() :: string() | non_neg_integer().

-type options() :: [
    inject
    | monitor
    | promiscuous
    | verbose
    | {buffer, arg_num()}
    | {chroot, string()}
    | {cluster_id, arg_num()}
    | {direction, in | out | inout}
    | {env, string()}
    | {exec, string()}
    | {file, string()}
    | {filter, string()}
    | {group, string()}
    | {interface, string()}
    | {progname, string()}
    | {snaplen, arg_num()}
    | {time_unit, time_unit()}
    | {timeout, arg_num() | infinity | immediate}
    | {user, string()}
    | {verbose, arg_num()}
].

-spec setopts(options(), options()) -> options().
setopts([], Options) ->
    proplists:compact(Options);
setopts([{Key, Val} | Rest], Options) ->
    case proplists:get_value(Key, Options) of
        undefined -> setopts(Rest, [{Key, Val} | Options]);
        _ -> setopts(Rest, Options)
    end.

%% @private
-spec getopts(options()) -> [string()].
getopts(Options) when is_list(Options) ->
    Exec = exec(Options),
    Progname = proplists:get_value(progname, Options, progname()),
    Cpu_affinity = cpu_affinity(Options),
    Filter = proplists:get_value(filter, Options, ""),
    Switches0 = lists:append([optarg(Opt) || Opt <- Options]),
    Switches = Switches0 ++ [Filter],
    [Cmd | Argv] = [
        N
     || N <- string:tokens(Exec, " ") ++ [Cpu_affinity, Progname | Switches],
        N /= ""
    ],
    [find_executable(Cmd) | Argv].

-spec optarg(atom() | tuple()) -> string().
optarg({buffer, Arg}) -> switch("b", maybe_string(Arg));
optarg({chroot, Arg}) -> switch("d", Arg);
optarg({cluster_id, Arg}) -> switch("e", env("PCAP_PF_RING_CLUSTER_ID", Arg));
optarg({env, Arg}) -> switch("e", Arg);
optarg({file, Arg}) -> switch("f", Arg);
optarg({group, Arg}) -> switch("g", Arg);
optarg({interface, Arg}) -> switch("i", Arg);
optarg(monitor) -> switch("M");
optarg(promiscuous) -> switch("P");
optarg({snaplen, Arg}) -> switch("s", maybe_string(Arg));
optarg({time_unit, Arg}) -> switch("T", time_unit(Arg));
optarg({timeout, immediate}) -> switch("t", "0");
optarg({timeout, infinity}) -> switch("t", "-1");
optarg({timeout, Arg}) -> switch("t", maybe_string(Arg));
optarg({user, Arg}) -> switch("u", Arg);
optarg(verbose) -> switch("v");
optarg({verbose, 0}) -> "";
optarg({verbose, Arg}) -> switch(string:copies("v", Arg));
optarg(inject) -> switch("X");
optarg({direction, Arg}) -> switch("Q", maybe_string(Arg));
optarg(_) -> "".

switch(Switch) -> [lists:concat(["-", Switch])].

switch(Switch, Arg) -> [lists:concat(["-", Switch]), Arg].

env(Key, Val) -> lists:concat([Key, "=", maybe_string(Val)]).

-spec basedir() -> string().
basedir() ->
    case code:priv_dir(?MODULE) of
        {error, bad_name} ->
            filename:join([filename:dirname(code:which(?MODULE)), "..", "priv", ?MODULE]);
        Dir ->
            Dir
    end.

-spec progname() -> string().
progname() -> filename:join([basedir(), ?MODULE]).

-spec exec(proplists:proplist()) -> string().
exec(Options) ->
    Exec = proplists:get_value(exec, Options, "sudo"),
    case proplists:is_defined(file, Options) of
        true -> "";
        false -> Exec
    end.

-spec cpu_affinity(proplists:proplist()) -> string().
cpu_affinity(Options) ->
    case proplists:get_value(cpu_affinity, Options) of
        undefined -> "";
        CPUs -> "taskset -c " ++ CPUs
    end.

-spec timeout() -> 0 | 500.
timeout() ->
    case os:type() of
        {unix, linux} -> 0;
        _ -> 500
    end.

%-spec time_unit(time_unit()) -> "0" | "1".
time_unit(timestamp) -> "0";
time_unit(microsecond) -> "1".

find_executable(Exe) ->
    case os:find_executable(Exe) of
        false -> erlang:error(badarg, [Exe]);
        N -> N
    end.

maybe_string(T) when is_list(T) -> T;
maybe_string(T) when is_integer(T) -> integer_to_list(T);
maybe_string(T) when is_atom(T) -> atom_to_list(T);
maybe_string(T) when is_binary(T) -> binary_to_list(T).

File: epcap_nif/c_src/epcap.c
/* Copyright (c) 2009-2013, Michael Santos <michael.santos@gmail.com>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * Redistributions of source code must retain the above copyright
 * notice, this list of conditions and the following disclaimer.
 *
 * Redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution.
 *
 * Neither the name of the author nor the names of its contributors
 * may be used to endorse or promote products derived from this software
 * without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
#include <ei.h>

#include "epcap.h"

int epcap_open(EPCAP_STATE *ep);
int epcap_init(EPCAP_STATE *ep);
void epcap_loop(EPCAP_STATE *ep);
void epcap_ctrl(const char *ctrl_evt);
void epcap_response(struct pcap_pkthdr *hdr, const u_char *pkt, unsigned int datalink);
void epcap_send_free(ei_x_buff *msg);
void epcap_watch();
void gotsig(int sig);
void usage(EPCAP_STATE *ep);

int child_exited = 0;

/* On some platforms (Linux), poll() (used by pcap)
 * will return EINVAL if RLIMIT_NOFILES < numfd */
#ifndef EPCAP_RLIMIT_NOFILES
#define EPCAP_RLIMIT_NOFILES 0
#warning "Using default value of EPCAP_RLIMIT_NOFILES=0"
#endif


    int
main(int argc, char *argv[])
{
    EPCAP_STATE *ep = NULL;
    pid_t pid = 0;
    int ch = 0;
    int fd = 0;


    IS_NULL(ep = calloc(1, sizeof(EPCAP_STATE)));

    ep->snaplen = SNAPLEN;
    ep->timeout = TIMEOUT;

    while ( (ch = getopt(argc, argv, "d:f:g:hi:MPs:t:u:v")) != -1) {
        switch (ch) {
            case 'd':   /* chroot directory */
                IS_NULL(ep->chroot = strdup(optarg));
                break;
            case 'f':
                IS_NULL(ep->file = strdup(optarg));
                ep->runasuser = 1;
                break;
            case 'g':
                IS_NULL(ep->group = strdup(optarg));
                break;
            case 'i':
                IS_NULL(ep->dev = strdup(optarg));
                break;
            case 'M':
                ep->rfmon = 1;
                break;
            case 'P':
                ep->promisc = 1;
                break;
            case 's':
                ep->snaplen = (size_t)atoi(optarg);
                break;
            case 't':
                ep->timeout = (u_int32_t)atoi(optarg);
                break;
            case 'u':
                IS_NULL(ep->user = strdup(optarg));
                break;
            case 'v':
                ep->verbose++;
                break;
            case 'h':
            default:
                usage(ep);
        }
    }

    argc -= optind;
    argv += optind;

    IS_NULL(ep->filt = strdup( (argc == 1) ? argv[0] : EPCAP_FILTER));

    IS_LTZERO(fd = open("/dev/null", O_RDWR));

    epcap_priv_issetuid(ep);
    IS_LTZERO(epcap_open(ep));
    if (epcap_priv_drop(ep) < 0)
        exit (1);

    signal(SIGCHLD, gotsig);

    switch (pid = fork()) {
        case -1:
            err(EXIT_FAILURE, "fork");
        case 0:
            IS_LTZERO(dup2(fd, STDIN_FILENO));
            IS_LTZERO(close(fd));
            IS_LTZERO(epcap_init(ep));
            IS_LTZERO(epcap_priv_rlimits(EPCAP_RLIMIT_NOFILES));
            epcap_loop(ep);
            break;
        default:
            if ( (dup2(fd, STDOUT_FILENO) < 0) ||
                (close(fd) < 0))
                goto CLEANUP;

            pcap_close(ep->p);

            if (epcap_priv_rlimits(0) < 0)
                goto CLEANUP;

            epcap_watch();

CLEANUP:
            (void)kill(pid, SIGTERM);
            break;
    }

    exit (0);
}


    void
epcap_watch()
{
    int fd = STDIN_FILENO;
    fd_set rfds;

    FD_ZERO(&rfds);
    FD_SET(fd, &rfds);

    if (child_exited == 0)
        (void)select(fd+1, &rfds, NULL, NULL, NULL);
}


    int
epcap_open(EPCAP_STATE *ep)
{
    char errbuf[PCAP_ERRBUF_SIZE];

    if (ep->file) {
        PCAP_ERRBUF(ep->p = pcap_open_offline(ep->file, errbuf));
    } else {
        if (ep->dev == NULL)
            PCAP_ERRBUF(ep->dev = pcap_lookupdev(errbuf));

        PCAP_ERRBUF(ep->p = pcap_open_live(ep->dev, ep->snaplen, ep->promisc, ep->timeout, errbuf));

        /* monitor mode */
#ifdef PCAP_ERROR_RFMON_NOTSUP
        if (pcap_can_set_rfmon(ep->p) == 1)
            (void)pcap_set_rfmon(ep->p, ep->rfmon);
#endif
    }

    return (0);
}


    int
epcap_init(EPCAP_STATE *ep)
{
    struct bpf_program fcode;
    char errbuf[PCAP_ERRBUF_SIZE];

    u_int32_t ipaddr = 0;
    u_int32_t ipmask = 0;


    if (pcap_lookupnet(ep->dev, &ipaddr, &ipmask, errbuf) == -1) {
        VERBOSE(1, "%s", errbuf);
        ipmask=PCAP_NETMASK_UNKNOWN;
    }

    VERBOSE(2, "[%s] Using filter: %s\n", __progname, ep->filt);

    if (pcap_compile(ep->p, &fcode, ep->filt, 1 /* optimize == true */, ipmask) != 0) {
        VERBOSE(1, "pcap_compile: %s", pcap_geterr(ep->p));
        return (-1);
    }

    if (pcap_setfilter(ep->p, &fcode) != 0) {
        VERBOSE(1, "pcap_setfilter: %s", pcap_geterr(ep->p));
        return (-1);
    }

    return (0);

}


    void
epcap_loop(EPCAP_STATE *ep)
{
    pcap_t *p = ep->p;
    struct pcap_pkthdr *hdr = NULL;
    const u_char *pkt = NULL;

    int read_packet = 1;
    int datalink = pcap_datalink(p);

    while (read_packet) {
        switch (pcap_next_ex(p, &hdr, &pkt)) {
            case 0:     /* timeout */
                VERBOSE(1, "timeout reading packet");
                break;
            case 1:     /* got packet */
                epcap_response(hdr, pkt, datalink);
                break;
            case -2:    /* eof */
                VERBOSE(1, "end of file");
                epcap_ctrl("eof");
                read_packet = 0;
                break;
            case -1:    /* error reading packet */
                VERBOSE(1, "%s", pcap_geterr(p));
                /* fall through */
            default:
                read_packet = 0;
        }
    }
}

void epcap_ctrl(const char *ctrl_evt)
{
    ei_x_buff msg;

    IS_FALSE(ei_x_new_with_version(&msg));
    IS_FALSE(ei_x_encode_tuple_header(&msg, 2));
    IS_FALSE(ei_x_encode_atom(&msg, "epcap"));
    IS_FALSE(ei_x_encode_atom(&msg, ctrl_evt));

    epcap_send_free(&msg);
}

    void
epcap_response(struct pcap_pkthdr *hdr, const u_char *pkt, unsigned int datalink)
{
    ei_x_buff msg;


    IS_FALSE(ei_x_new_with_version(&msg));

    /* {packet, DatalinkType, Time, ActualLength, Packet} */
    IS_FALSE(ei_x_encode_tuple_header(&msg, 5));
    IS_FALSE(ei_x_encode_atom(&msg, "packet"));

    /* DataLinkType */
    IS_FALSE(ei_x_encode_long(&msg, datalink));

    /* {MegaSec, Sec, MicroSec} */
    IS_FALSE(ei_x_encode_tuple_header(&msg, 3));
    IS_FALSE(ei_x_encode_long(&msg, abs(hdr->ts.tv_sec / 1000000)));
    IS_FALSE(ei_x_encode_long(&msg, hdr->ts.tv_sec % 1000000));
    IS_FALSE(ei_x_encode_long(&msg, hdr->ts.tv_usec));

    /* ActualLength} */
    IS_FALSE(ei_x_encode_long(&msg, hdr->len));

    /* Packet */
    IS_FALSE(ei_x_encode_binary(&msg, pkt, hdr->caplen));

    /* } */

    epcap_send_free(&msg);
}

void epcap_send_free(ei_x_buff *msg)
{
    u_int16_t len = 0;

    len = htons(msg->index);
    if (write(STDOUT_FILENO, &len, sizeof(len)) != sizeof(len))
        errx(EXIT_FAILURE, "write header failed");

    if (write(STDOUT_FILENO, msg->buff, msg->index) != msg->index)
        errx(EXIT_FAILURE, "write packet failed: %d", msg->index);

    ei_x_free(msg);
}

    void
gotsig(int sig)
{
    switch (sig) {
        case SIGCHLD:
            child_exited = 1;
            break;
        default:
            break;
    }
}

    void
usage(EPCAP_STATE *ep)
{
    (void)fprintf(stderr, "%s, %s\n", __progname, EPCAP_VERSION);
    (void)fprintf(stderr,
            "usage: %s <options>\n"
            "              -d <directory>   chroot directory\n"
            "              -i <interface>   interface to snoop\n"
            "              -f <filename>    read from file instead of live capture\n"
#ifdef PCAP_ERROR_RFMON_NOTSUP
            "              -M               wireless monitor (rfmon) mode\n"
#endif
            "              -P               promiscuous mode\n"
            "              -g <group>       unprivileged group\n"
            "              -u <user>        unprivileged user\n"
            "              -s <length>      packet capture length\n"
            "              -t <millisecond> capture timeout\n"
            "              -v               verbose mode\n",
            __progname
            );

    exit (EXIT_FAILURE);
}
